File: solutions/fable-1/fable/perl/roman.pl
use strict;
use warnings;

my @map = (
    [1000, 'M'], [900, 'CM'], [500, 'D'], [400, 'CD'],
    [100, 'C'], [90, 'XC'], [50, 'L'], [40, 'XL'],
    [10, 'X'], [9, 'IX'], [5, 'V'], [4, 'IV'], [1, 'I'],
);

while (my $line = <STDIN>) {
    $line =~ s/^\s+|\s+$//g;
    next if $line eq '';
    my $n = int($line);
    my $out = '';
    for my $pair (@map) {
        my ($v, $s) = @$pair;
        while ($n >= $v) {
            $out .= $s;
            $n -= $v;
        }
    }
    print "$out\n";
}

File: solutions/fable-1/fable/ruby/roman.rb
PAIRS = [
  [1000, "M"], [900, "CM"], [500, "D"], [400, "CD"],
  [100, "C"], [90, "XC"], [50, "L"], [40, "XL"],
  [10, "X"], [9, "IX"], [5, "V"], [4, "IV"], [1, "I"]
].freeze

def to_roman(n)
  out = +""
  PAIRS.each do |value, sym|
    while n >= value
      out << sym
      n -= value
    end
  end
  out
end

$stdin.each_line do |line|
  s = line.strip
  next if s.empty?
  puts to_roman(s.to_i)
end
